Add CG

vladimir-ch opened the pull request:
A proposal for a possible implementantion of the non-linear conjugate gradient method.

Points to consider:
* CG variants are quite verbose at the moment. We could consider renaming FletcherReeves to FR, PolakRibierePolyak to PRPPlus, etc.
* More variants can be added easily
* After a very limited testing, surprisingly, Hestenes-Stiefel seems to perform better than Polak-Ribiere.
* Restarting after n iterations without a restart and after an angle condition fails is performed always. A user setting to turn these restarting strategies on/off could be added.

#16 should be reviewed and merged first.

btracey commented on cg.go:
```diff
@@ -0,0 +1,162 @@
+package optimize
+
```
All these exported structs need comments. I'm not familiar with CGVariant, so I personally would appreciate some explanation. Up to you what the appropriate level is, as we obviously aren't going to write a book in the comments.

btracey commented on cg.go:
```diff
@@ -0,0 +1,162 @@
+	angleRestartTreshold   = 0.9
+)
+
+type CGVariant interface {
```
I don't particularly like the word Variant. I understand what you mean, but it's a pretty vague term. Wikipedia calls it a "Preconditioner" which is much more descriptive. If you don't think that's misleading, I'd rather see CGPreconditioner (or CGPrecon) than CGVariant.

vladimir-ch replied:
CGVariant computes the scaling parameter in the direction update formula, which is what differentiates various CG methods one from another. Therefore I like CGVariant, alternatively it could be called CGFormula. CGPreconditioner seems out of place here.

btracey commented on cg.go:
```diff
@@ -0,0 +1,162 @@
+	copy(direction, loc.Gradient)
+	floats.Scale(-1, direction)
+
+	var β float64
```
@kortschak what is our policy on Unicode? I would probably just write beta.

kortschak replied:
Generally people frown on non-ASCII named vars. I don't think we should use them.

vladimir-ch replied:
I admit that that β variable was just my fancy. With a slight disappointment I will change it to plain beta.

btracey commented on cg.go:
```diff
@@ -0,0 +1,162 @@
+
+type CGVariant interface {
+	Init(loc Location)
+	Beta(gradient, gradPrev, dirPrev []float64) float64
```
Looking at the CG code, there should definitely be a comment on Beta specifying that zero is a special value.

btracey commented on cg.go:
```diff
@@ -0,0 +1,162 @@
+		β = cg.Variant.Beta(loc.Gradient, cg.gradPrev, cg.dirPrev)
+	}
+	if β != 0 {
+		floats.AddScaled(direction, β, cg.dirPrev)
```
Could you add a quick high-level comment here.

// If beta == 0, it means that it is time to restart (because of iteration number or the CGVariant). When we restart, ...

btracey replied:
Sorry, the beginning to my comment was totally wrong. Instead it should be something like 
// We have a scaling factor from the variant. Update the direction from the CG rule
(and if it's an easy rule in math notation, say it).

vladimir-ch replied:
I basically rewrote NextDirection and added implementation comments explaining the function in detail.

btracey commented on cg.go:
```diff
@@ -0,0 +1,162 @@
+			cg.iterFromRestart = 0
+		}
+	} else {
+		cg.iterFromRestart = 0
```
And then here.
// It is time to restart. When we restart we take the true gradient as the direction.
(at least, that's what it looks like to me)

btracey commented:
There need to be tests along with these additions.

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+	angleRestartThreshold  = 0.9
+)
+
+// CGVariant represents the formula for computing the scaling parameter (in
```
"CGVariant calculates the scaling parameter beta" (With greek symbol)

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+// CG implements the nonlinear conjugate gradient method for solving nonlinear
+// unconstrained optimization problems [1]. It is a line search method that
+// generates the search directions d_k according to the formula
+// d_{k+1} = -∇f_{k+1} + β_k*d_k,   d_0 = -∇f_0.
```
Space needed for godoc

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+// unconstrained optimization problems [1]. It is a line search method that
+// generates the search directions d_k according to the formula
+// d_{k+1} = -∇f_{k+1} + β_k*d_k,   d_0 = -∇f_0.
+// Different variants of the conjugate gradient method differ in the choice
```
"Variants of ... differ", no need to repeat

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+// d_{k+1} = -∇f_{k+1} + β_k*d_k,   d_0 = -∇f_0.
+// Different variants of the conjugate gradient method differ in the choice
+// of the parameter β_k. Conjugate gradient method requires no matrix storage
+// and is faster than the gradient descent method, but L-BFGS is usually more
```
"Faster" and "more efficient" are misleading, because update flops is one metric,  and function evaluations are a totally different one.  Instead,  just say that CG usually requires fewer function evaluations,  and leave off words that sound like speed. I'm not sure what you mean by lbfgs being more robust, I've heard CG is more robust when the gradients are noisy

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+//
+// CG implements a restart strategy that takes the steepest descent direction
+// (i.e., d_{k+1} = -∇f_{k+1}) whenever any of the following conditions holds.
+// * A certain number of iterations has elapsed without a restart. This number
```
Again,  spaces are needed for godoc. Make sure you verify that asterisks are okay for bullets (don't have access atm)

jonlawlor replied:
They come through as asterisks just fine.  Also, the ArmijoConditionMet function uses a different format for bullet points ("-" instead of "*") and also has incorrect spacing in its fixed width section.  Several of the comments have different fixed width formatting sections.

You should be able to see it at http://godoc.org/github.com/jonlawlor/optimize#CG

btracey replied:
@jonlawlor  is right, we should keep with the "-". Why is the Armijio spacing wrong? I observe visually that it is, but the actual comment looks basically identical to that of StrongWolfeConditionsMet which shows up in godoc just fine.

jonlawlor replied:
I think https://github.com/gonum/optimize/blob/master/linesearch.go#L118 has an extra tab.

btracey replied:
Weird ... why don't I see that locally?

````
brendan:~/Documents/mygo/src/github.com/gonum/optimize$ git branch
  bfgsinit
  decreasetol
* master
brendan:~/Documents/mygo/src/github.com/gonum/optimize$ git status
On branch master
Your branch is up-to-date with 'origin/master'.

nothing to commit, working directory clean
brendan:~/Documents/mygo/src/github.com/gonum/optimize$ git pull
Already up-to-date.
````

jonlawlor replied:
Maybe your text editor is hiding it?  Try
```bash
cat linesearch.go | grep -m 1 initGrad
// 	- initGrad < 0
```

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+// CG implements a restart strategy that takes the steepest descent direction
+// (i.e., d_{k+1} = -∇f_{k+1}) whenever any of the following conditions holds.
+// * A certain number of iterations has elapsed without a restart. This number
+//   depends on the dimension of the problem and is controllable via
```
Say "This number is controllable via IterationRestartFactor and if 0 is set to a reasonable default based on the problem dimension"

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+// The line search for CG must yield step sizes that satisfy the strong Wolfe
+// conditions at every iteration, otherwise the generated search direction
+// might fail to be a descent direction. The line search should be more
+// accurate compared to the case of Newton-like methods, which can be achieved
```
s/accurate/stringent
compared with those for Newton-like methods

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+//     methods. Pacific Journal of Optimization, 2 (2006), pp. 35-58.
+type CG struct {
+	// LinesearchMethod must satisfy the strong Wolfe conditions at every
+	// iteration. If LinesearchMethod == nil, an appropriate default is chosen
```
Just say ", an appropriate default is chosen".

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+	// whenever ceil(IterationRestartFactor*(dimension of the problem))
+	// iterations have elapsed without a restart.
+	// For medium and large-scale problems IterationRestartFactor should be set
+	// to 1 and greater values should be used for low-dimensional problems.
```
to 1, for low-dimensional problems a larger value should be chosen.

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+	// iterations have elapsed without a restart.
+	// For medium and large-scale problems IterationRestartFactor should be set
+	// to 1 and greater values should be used for low-dimensional problems.
+	// Care must be taken because if the resulting number is 1, the iteration
```
Note that if IterationRestartFactor == 1, CG becomes identical to gradient descent.

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+	// the steepest descent method.
+	// If IterationRestartFactor is 0, it will be set to 6.
+	IterationRestartFactor float64
+	// The method is restarted whenever the gradients at two consecutive
```
AngleRestartThreshold sets the threshold angle for restart. If the angle between two consecutive gradients is greater than AngleRestartThreshold
  formula
the decent direction is the same as the gradient. A value of AngleRestartThreshold closer to 1 will reduce the number of restarts.

btracey replied:
We also need to note that AngleRestartThreshold is the cosine of the angle, not the angle iteslf

btracey replied:
Err, cos of the angle.

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+func (cg *CG) Init(loc Location, funcInfo *FunctionInfo, xNext []float64) (EvaluationType, IterationType, error) {
+	if cg.LinesearchMethod == nil {
```
Should there be panics for Angle / IterationRestart being less than zero?

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+	var restart bool
+	// Check if the method should be restarted because too many iterations have
+	// been taken without a restart.
+	if cg.iterFromRestart%cg.restartAfter == 0 {
```
Can we just make it a comparison rather than a mod?

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+		restart = true
+	}
+	// Check if the method should be restarted because the last two gradients
+	// point in almost opposite directions.
```
almost orthogonal directions. (based on my understanding of the function)

vladimir-ch replied:
It would be *far from* orthogonal directions but only if the dot product were in an absolute value. As it is now, it indeed restarts as it is written, that is, when the gradients point in almost opposite directions. I chose the non-absolute-value version as it seems to assure convergence in the same number of cases with less restarts.

btracey replied:
Yea, you're right, I got confused with the negative sign. My fault.

Would it make more sense to have the restart happen if the angle is less than AngleThreshhold and remove the negative sign? In this case, AngleThreshhold would default to -0.9. This way it aligns with the Cos formula, 1 and they're perfectly aligned, 0 they're othogonal, -1 they're exactly opposite.

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+	// Check if the method should be restarted because the last two gradients
+	// point in almost opposite directions.
+	gNorm := floats.Norm(loc.Gradient, 2)
+	gDotGPrev := floats.Dot(loc.Gradient, cg.gradPrev)
```
can we just call this gDot?

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+
+// PolakRibierePolyak implements the Polak-Ribiere-Polyak variant of the CG
+// method that computes the scaling parameter β_k according to the formula
+// β_k = max(0, ∇f_{k+1}·y_k / |∇f_k|^2),
```
Formula should be indented with a space so it gets its formatting box. This should happen with all of the variants for both of the formulas.

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+
+func (fr *FletcherReeves) Beta(grad, gradPrev, dirPrev []float64) (beta float64) {
+	gNorm := floats.Norm(grad, 2)
+	beta = gNorm * gNorm / fr.gPrevNorm / fr.gPrevNorm
```
Please make it
````
beta = (gNorm / fr.gPrevNorm) * (gNorm / gPrevNorm)
````
Should be better behaved in edge cases.

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+// computes the scaling parameter β_k according to the formula
+// β_k = |∇f_{k+1}|^2 / |∇f_k|^2.
+type FletcherReeves struct {
+	gPrevNorm float64
```
Just call it prevNorm. There's no other norm or other variables involved at all

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+func (fr *FletcherReeves) Beta(grad, gradPrev, dirPrev []float64) (beta float64) {
+	gNorm := floats.Norm(grad, 2)
```
Just call it norm (see above)

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+// β_k = max(0, ∇f_{k+1}·y_k / |∇f_k|^2),
+// where y_k = ∇f_{k+1} - ∇f_k.
+type PolakRibierePolyak struct {
+	gPrevNorm float64
```
Same comments about norm

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+
+func (pr *PolakRibierePolyak) Beta(grad, gradPrev, dirPrev []float64) (beta float64) {
+	gNorm := floats.Norm(grad, 2)
+	gDotGPrev := floats.Dot(grad, gradPrev)
```
You can call this "dot"

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+func (pr *PolakRibierePolyak) Beta(grad, gradPrev, dirPrev []float64) (beta float64) {
+	gNorm := floats.Norm(grad, 2)
+	gDotGPrev := floats.Dot(grad, gradPrev)
+	beta = (gNorm*gNorm - gDotGPrev) / pr.gPrevNorm / pr.gPrevNorm
```
This is probably better separated out too
beta = (norm / pr.prevNorm) * (norm / pr.prevNorm) - dot / (pr.prevNorm * pr.prevNorm)

btracey replied:
Actually, I take it back. probably best as
beta = (norm * norm - dot) / (pr.prevNorm * pr.prevNorm)

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+
+func (hz *HagerZhang) Beta(grad, gradPrev, dirPrev []float64) (beta float64) {
+	floats.SubTo(hz.y, grad, gradPrev)
+	dDotY := floats.Dot(dirPrev, hz.y)
```
dirDotY

btracey commented on cg.go:
```diff
@@ -0,0 +1,293 @@
+	floats.SubTo(hz.y, grad, gradPrev)
+	dDotY := floats.Dot(dirPrev, hz.y)
+	gDotY := floats.Dot(grad, hz.y)
+	gDotD := floats.Dot(grad, dirPrev)
```
gDotDir

jonlawlor commented:
I believe https://github.com/gonum/optimize/blob/master/unconstrained_test.go#L11 needs to be changed.

vladimir-ch commented:
I think I addressed all your comments, @btracey , could you PTAL? I am not sure if the comments for AngleRestartFactor and IterationsRestartFactor are clear enough. CG panics if AngleRestartFactor is not in [-1,0], because that is the only reasonable interval of values for it, but it too can be reconsidered.

vladimir-ch commented:
Also, I will add tests later after reviewing and merging change-tests branch.

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+//     is controllable via IterationRestartFactor and if equal to 0, it is set to
+//     a reasonable default based on the problem dimension.
+//   - The angle between the gradients at two consecutive iterations ∇f_k and
+//     ∇f_{k+1} does not satisfy a certain criterion.
```
instead of "does not satisfy a certain criterion" can we just say "is too large"?

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+//     a reasonable default based on the problem dimension.
+//   - The angle between the gradients at two consecutive iterations ∇f_k and
+//     ∇f_{k+1} does not satisfy a certain criterion.
+//   - β_k given by Variant.Beta() is equal to zero.
```
CGVaraint.Beta()

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+	// AngleRestartThreshold sets the threshold angle for restart. If the
+	// cosine of the angle between two consecutive gradients is smaller than
+	// AngleRestartThreshold, that is, if
+	//
```
The extra newlines aren't necessary.

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+	//
+	//   ∇f_k·∇f_{k+1} / (|∇f_k| |∇f_{k+1}|) < AngleRestartThreshold,
+	//
+	// the descent direction is the same as the negative of the gradient.
```
What does this comment have to do with anything? The descent direction is not mentioned anywhere else.

I think it's fine to just delete this sentence and end the thought at the comment.

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+	//   ∇f_k·∇f_{k+1} / (|∇f_k| |∇f_{k+1}|) < AngleRestartThreshold,
+	//
+	// the descent direction is the same as the negative of the gradient.
+	// A value of AngleRestartThreshold closer to -1 will reduce the number of
```
"closer to -1 (successive gradients in exact opposite directions) will tend to reduce"

btracey commented:
I had a couple of minor comments, but nice job on the documentation.

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+	// the last two gradients is too large.
+	gDot := floats.Dot(loc.Gradient, cg.gradPrev)
+	gNorm := floats.Norm(loc.Gradient, 2)
+	cos := math.Max(-1, gDot/(gNorm*cg.gradPrevNorm))
```
What does the Max do? It seems to me like the only effect is has is to allow AngleRestartThreshold == -1 to act as "never restart because of the angle" even when the gradients are really poorly scaled. If the line below this was "<=" instead of less than, the max would have no effect, because cg.AngleRestartThreshold >= -1.

vladimir-ch replied:
Yes, allowing AngleRestartThreshold == -1 to skip the angle test was the intention. Given that CG easily fails without this condition, it make no sense to try to reserve -1 for skipping it. I modified the condition (and the docs) back to <= and removed the Max.

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+	}
+
+	// Compute the scaling factor β_k according to the given CG variant. Do
+	// this always, because cg.Variant may be keeping its inner state that
```
s/alwals/even when restarting

cg.Variant may be keeping _an_ inner state  (underscores mine to highlight the change, don't actually include them)

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+
+// FletcherReeves implements the Fletcher-Reeves variant of the CG method that
+// computes the scaling parameter β_k according to the formula
+//
```
extra line not needed. This is true for all below.

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+// FletcherReeves implements the Fletcher-Reeves variant of the CG method that
+// computes the scaling parameter β_k according to the formula
+//
+//   β_k = |∇f_{k+1}|^2 / |∇f_k|^2.
```
You only need one extra space. (below as well)

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+	norm := floats.Norm(grad, 2)
+	dot := floats.Dot(grad, gradPrev)
+	beta = (norm*norm - dot) / (pr.prevNorm * pr.prevNorm)
+
```
remove newline.

btracey commented on cg.go:
```diff
@@ -0,0 +1,317 @@
+func (fr *FletcherReeves) Beta(grad, gradPrev, dirPrev []float64) (beta float64) {
+	norm := floats.Norm(grad, 2)
+	beta = (norm / fr.prevNorm) * (norm / fr.prevNorm)
+
```
remove newline

btracey commented:
LGTM when the comments are addressed. The only comment that addresses code function is the Max question. The rest are all documentation / minor style.

Thanks for working with me. This all looks really nice now.

vladimir-ch commented:
Thanks for reviewing the code, @btracey . Writing a compact, clear Go style documentation is hard, your comments have really helped me to improve it (not to mention the code itself).
